apps/system/js/app_install_manager.js

```diff
@@ -256,6 +261,56 @@
         this.authorUrl.textContent = '';
       }
 
+      // Warn about system level addons?
+      var warningEl = document.getElementById('system-addon-warning');
+      var showWarning = false;
+
+      if (manifest.content_scripts) {
+        var contentScripts = manifest.content_scripts;
+        var patterns = contentScripts.map(contentScript => {
+          return {
+            matches: new MatchPattern(contentScript.matches),
+            excludeMatches: new MatchPattern(contentScript.exclude_matches)
+          };
+        });
+
+        appLoop:
+        for (var appKey in applications.installedApps) {
+          if (!appKey) {
+            continue;
+          }
+
+          var installedApp = applications.installedApps[appKey];
+          var launchPath = installedApp.manifest.launch_path || '';
+
+          if (!installedApp.manifest.origin) {
+            continue;
+          }
+
+          var launchURL = new URL(launchPath, installedApp.manifest.origin);
+
+          for(var i = 0; i < patterns.length; i++) {
+            var pattern = patterns[i];
+            if (!pattern.excludeMatches.matches(launchURL) &&
+                pattern.matches.matches(launchURL)) {
+
+              if (installedApp.manifest.role === 'system' ||
+                  installedApp.manifest.role === 'homescreen') {
```

Comment:
Let's move this to outside the pattern matching for loop since we can know immediately if the apps are homscreen or system.